hoomd/md/methods.py

```diff
@@ -26,6 +26,60 @@ class _Method(_HOOMDBaseObject):
         Users should use the subclasses and not instantiate `_Method` directly.
     """
 
+class _MethodRATTLE(_Method):
+    """Base class rattle integration method.
+
+    Provides common methods for all subclasses which are compatible with the RATTLE
+    algorithm. Particles can be constrained to a manifold surface using the
+    RATTLE scheme. For poor initial conditions that include overlapping atoms,
+    a limit can be specified to the movement a particle is allowed to make in
+    one time step. After a few thousand time steps with the limit set, the
+    system should be in a safe state to continue with unconstrained integration.
+
+    Warning:
+        The particles should be initialised close to the implicit surface of
+        the manifold. Even though the particles are mapped to the set surface
+        automatically, the mapping can lead to small inter-particle distances and,
+        hence, large forces between particles!
+
+    For the MTK equations of motion, see:
+
+    * `S. Paquay and R. Kusters  2016 <https://doi.org/10.1016/j.bpj.2016.02.017>`_
+
+    Note:
+        Users should use the subclasses and not instantiate `_MethodRATTLE` directly.
+
+    """
+    def __init__(self, manifold_constraint, eta):
+        param_dict = ParameterDict(
+            manifold_constraint = OnlyTypes(Manifold, allow_none=True),
+            eta=float(eta)
+            )
+        param_dict.update(dict(manifold_constraint=manifold_constraint))
+        # set defaults
+        self._param_dict.update(param_dict)
+
+
+    def _attach_constraint(self,sim):
+        if not self.manifold_constraint._added:
+            self.manifold_constraint._add(sim)
+        else:
+            if sim != self.manifold_constraint._simulation:
+                raise RuntimeError("{} object's manifold_constraint is used in a "
+                                    "different simulation.".format(type(self)))
+        if not self.manifold_constraint._attached:
+            self.manifold_constraint._attach()
+
+    def _getattr_param(self, attr):
+        if self._attached:
+            parameter = getattr(self._cpp_obj, attr)
+            if attr == "manifold_constraint":
```

Comment:
```suggestion
            # "manifold_constraint" returns a dict rather than an Manifold object
            # so we convert it before returning it to the user.
            if attr == "manifold_constraint":
```

If `manifold` objects are immutable (as I have argued they should be given their usage), We don't even need this something like

```python
def _getattr_param(self, attr):
    if self._attached:
        if attr == "manifold_constraint":
            return self._param_dict[attr]
        return getattr(self._cpp_obj, attr)
    return self._param_dict[attr]